### PR TITLE
Copied typings from DefinitelyTyped

### LINKED ISSUE #10: Add TypeScript (@types) typings
Spinning off of https://github.com/chaijs/chai/pull/1092#discussion_r152858013...

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
Adds TypeScript as a dev dependency and specifies `index.d.ts` as typings, with `--noEmit` tests run as part of `test`.

Fixes #10.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,27 +1,29 @@
 {
-    "name": "assertion-error"
```
Comment (lucasfcosta) [inline package.json]: It feels so good to have this file indented like this 😆

**Comment by lucasfcosta** [comment]
Thanks @JoshuaKGoldberg!

This looks awesome! LGTM.

Seems like we will have TypeDefinitions on our next release 😄

**Comment by demurgos** [comment]
It's a good start.
It would be nice to capture the relation between the `props` argument the property available on the instances.

**Comment by demurgos** [comment]
Here is a stricter modelization supporting the property types (I'm using an internal module declaration style because I was testing in my local project):

```typescript
declare module "assertion-error" {
  type AssertionError<T> = Error & T & {showDiff: boolean};

  interface AssertionErrorConstructor {
    new<T>(message: string, props?: T, ssf?: Function): AssertionError<T>;
  }

  const AssertionError: AssertionErrorConstructor;

  export = AssertionError;
}

```

You can use it as:
```typescript
const assertionError: AssertionError<{foo: number}> = new AssertionError("msg", {foo: 42});
const msg: string = assertionError.message;
const foo: number = assertionError.foo;
```

File: package.json (hunk 1 above)
Comment (JoshuaKGoldberg) [reply package.json]: Yeah this is what happens automatically. I can't stand pre-line commas anyway! D:

**Comment by JoshuaKGoldberg** [comment]
@demurgos added! I defaulted `T` to `= {}` because there might be usages of AssertionError out there that don't provide an explicit generic.

**Comment by demurgos** [comment]
Thank you for taking time to update the typings.
It looks pretty good. I also added your branch as a dependency of one of my projects and experimented with it: everything works fine. No conflicts, support for require imports, ES imports, type inference for the properties, autocompletion. Good work.

@keithamus 
I'd recommend to merge it and publish it to npm so we can move forward with https://github.com/chaijs/chai/pull/1092.

**Comment by demurgos** [comment]
Could we move forward with this issue?

**Comment by keithamus** [comment]
Does someone want to raise a PR bumping the package.json so we can release?

**Comment by demurgos** [comment]
I'll do a PR.
I think it should be semver minor bump because it's a feature for Typescript, but you may also consider it to be a patch because it has no runtime impact.

**Pull request merged**